tests/analyser/analyser.cpp

```diff
@@ -0,0 +1,362 @@
+
+    EXPECT_EQ(libcellml::AnalyserModel::Type::INVALID, analyser->model()->type());
+}
+
+TEST(Analyser, nonInitialisedState)
+{
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("analyser/non_initialised_state.cellml"));
+
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    const std::vector<std::string> expectedIssues = {
+        "Variable 'x' in component 'my_component' of model 'non_initialised_state' is used in an ODE, but it is not initialised.",
+    };
+    const std::vector<libcellml::Issue::Cause> expectedCauses = {
+        libcellml::Issue::Cause::ANALYSER,
+    };
+
+    auto analyser = libcellml::Analyser::create();
+
+    analyser->processModel(model);
+
+    EXPECT_EQ_ISSUES_CAUSES(expectedIssues, expectedCauses, analyser);
+
+    EXPECT_EQ(libcellml::AnalyserModel::Type::UNDERCONSTRAINED, analyser->model()->type());
+}
+
+TEST(Analyser, underconstrained)
+{
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("analyser/underconstrained.cellml"));
+
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    const std::vector<std::string> expectedIssues = {
+        "Variable 'x' in component 'my_component' of model 'my_model' is not computed.",
+    };
+    const std::vector<libcellml::Issue::Cause> expectedCauses = {
+        libcellml::Issue::Cause::ANALYSER,
+    };
+
+    auto analyser = libcellml::Analyser::create();
+
+    analyser->processModel(model);
+
+    EXPECT_EQ_ISSUES_CAUSES(expectedIssues, expectedCauses, analyser);
+
+    EXPECT_EQ(libcellml::AnalyserModel::Type::UNDERCONSTRAINED, analyser->model()->type());
+}
+
+TEST(Analyser, overconstrained)
+{
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("analyser/overconstrained.cellml"));
+
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    const std::vector<std::string> expectedIssues = {
+        "Variable 'x' in component 'my_component' of model 'my_model' is computed more than once.",
+    };
+    const std::vector<libcellml::Issue::Cause> expectedCauses = {
+        libcellml::Issue::Cause::ANALYSER,
+    };
+
+    auto analyser = libcellml::Analyser::create();
+
+    analyser->processModel(model);
+
+    EXPECT_EQ_ISSUES_CAUSES(expectedIssues, expectedCauses, analyser);
+
+    EXPECT_EQ(libcellml::AnalyserModel::Type::OVERCONSTRAINED, analyser->model()->type());
+}
+
+TEST(Analyser, unsuitablyConstrained)
+{
+    auto parser = libcellml::Parser::create();
+    auto model = parser->parseModel(fileContents("analyser/unsuitably_constrained.cellml"));
+
+    EXPECT_EQ(size_t(0), parser->issueCount());
+
+    const std::vector<std::string> expectedIssues = {
+        "Variable 'x' in component 'my_component' of model 'my_model' is not computed.",
+        "Variable 'y' in component 'my_component' of model 'my_model' is computed more than once.",
+    };
```

Comment:
see comment above.

Reply:
See https://github.com/cellml/libcellml/pull/631#discussion_r448764600.

Reply:
I'm still not happy with this.  How about "variable 'y' in component 'my_component' has more than one definition".  I think that "over-constrained" is the better terminology as it relates to the system of equations and is fairly standard.  Computed more than once means to me that you're iterating ;) which sounds fine!